packages/eslint-plugin/src/rules/parameter-properties.ts

```diff
@@ -0,0 +1,239 @@
+                  (node.parameter.left as TSESTree.Identifier).name;
+
+            context.report({
+              node,
+              messageId: 'preferClassProperty',
+              data: {
+                parameter: name,
+              },
+            });
+          }
+        },
+      };
+    }
+
+    interface PropertyNodes {
+      classProperty?: TSESTree.PropertyDefinition;
+      constructorAssignment?: TSESTree.AssignmentExpression;
+      constructorParameter?: TSESTree.Identifier;
+    }
+
+    const propertyNodesByNameStack: Map<string, PropertyNodes>[] = [];
+
+    function getNodesByName(name: string): PropertyNodes {
+      const propertyNodesByName =
+        propertyNodesByNameStack[propertyNodesByNameStack.length - 1];
+      const existing = propertyNodesByName.get(name);
+      if (existing) {
+        return existing;
+      }
+
+      const created: PropertyNodes = {};
+      propertyNodesByName.set(name, created);
+      return created;
+    }
+
+    const sourceCode = context.getSourceCode();
+
+    function typeAnnotationsMatch(
+      classProperty: TSESTree.PropertyDefinition,
+      constructorParameter: TSESTree.Identifier,
+    ): boolean {
+      if (
+        !classProperty.typeAnnotation ||
+        !constructorParameter.typeAnnotation
+      ) {
+        return (
+          classProperty.typeAnnotation === constructorParameter.typeAnnotation
+        );
+      }
+
+      return (
+        sourceCode.getText(classProperty.typeAnnotation) ===
+        sourceCode.getText(constructorParameter.typeAnnotation)
+      );
+    }
+
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        propertyNodesByNameStack.push(new Map());
+      },
+
+      ':matches(ClassDeclaration, ClassExpression):exit'(): void {
+        const propertyNodesByName = propertyNodesByNameStack.pop()!;
+
+        for (const [name, nodes] of propertyNodesByName) {
+          if (
+            nodes.classProperty &&
+            nodes.constructorAssignment &&
+            nodes.constructorParameter &&
+            typeAnnotationsMatch(
+              nodes.classProperty,
+              nodes.constructorParameter,
+            )
+          ) {
+            context.report({
+              data: {
+                parameter: name,
+              },
+              messageId: 'preferParameterProperty',
+              node: nodes.classProperty,
+            });
+          }
+        }
+      },
+
+      ClassBody(node): void {
+        for (const element of node.body) {
+          if (
+            element.type === AST_NODE_TYPES.PropertyDefinition &&
+            element.key.type === AST_NODE_TYPES.Identifier &&
+            !element.value &&
+            !allow.includes(getModifiers(element))
+          ) {
+            getNodesByName(element.key.name).classProperty = element;
+          }
+        }
+      },
+
+      'MethodDefinition[kind="constructor"]'(
```

Comment:
side note: we should add these sorts of things to our types so that we don't need to annotate manually every time.

Reply:
_Yes!_ I'd love to tackle #4065 once I have ... much more time 😄